Modules/_datetimemodule.c

```diff
@@ -663,6 +663,135 @@ set_date_fields(PyDateTime_Date *self, int y, int m, int d)
+        *var *= 10;
+        *var += (signed int)tmp;
+    }
+
+    return ptr;
+}
+
+static int parse_isoformat_date(const char *dtstr,
+                                int* year, int *month, int* day) {
+    /* Parse the date components of the result of date.isoformat()
+    *
+    *  Return codes:
+    *       0:  Success
+    *      -1:  Failed to parse date component
+    *      -2:  Failed to parse dateseparator
+    */
+    const char *p = dtstr;
+    p = parse_digits(p, year, 4);
+    if (NULL == p) { return -1; }
+    
+    if (*(p++) != '-') { return -2; }
+
+    p = parse_digits(p, month, 2);
+    if (NULL == p) { return -1; }
+
+    if (*(p++) != '-') { return -2; }
+
+    p = parse_digits(p, day, 2);
+    if (p == NULL) { return -1; }
+
+    return 0;
+}
+
+// Macro that short-circuits to timezone parsing
+#define PARSE_ISOFORMAT_ADVANCE_TIME_SEP(SEP) {       \
+    if (*p == '\0') { return 0; }                   \
+    switch(*(p++)) {                                \
+        case SEP:                                   \
+            break;                                  \
+        case '-':                                   \
+            tzsign = -1;                            \
+        case '+':                                   \
+            goto parse_timezone;                    \
+        default:                                    \
+            return -4;                              \
+    }                                               \
+}
+
+static int
+parse_isoformat_time(const char *dtstr, int* hour, int *minute, int *second,
+                     int* microsecond, int *tzoffset) {
+    // Parse the time portion of a datetime.isoformat() string
+    //
+    // Return codes:
+    //      0:  Success (no tzoffset)
+    //      1:  Success (with tzoffset)
+    //     -3:  Failed to parse time component
+    //     -4:  Failed to parse time separator
+    //     -5:  Malformed timezone string
+    //     -6:  Extra information at end of string
+
+    const char *p = dtstr;
+    
+    // Parse time - all components are optional except hour
+    int tzsign = 1;
+    p = parse_digits(p, hour, 2);
+    if (NULL == p) { return -3; }
+
+    PARSE_ISOFORMAT_ADVANCE_TIME_SEP(':');
+
+    p = parse_digits(p, minute, 2);
+    if (NULL == p) { return -3; }
+
+    PARSE_ISOFORMAT_ADVANCE_TIME_SEP(':');
+
+    p = parse_digits(p, second, 2);
+    if (NULL == p) { return -3; }
+
+    PARSE_ISOFORMAT_ADVANCE_TIME_SEP('.');
+
+    p = parse_digits(p, microsecond, 3);
+    if (NULL == p) { return -3; }
+
+    // Microseconds
+    if (*p >= '0' && *p <= '9') {
+        p = parse_digits(p, microsecond, 3);
+        if (NULL == p) { return -3; }
+    } else {
+        *microsecond *= 1000;
+    }
+
+    if (*p == '\0') { return 0; }
+    
+    switch(*(p++)) {
+        case '-':
+            tzsign = -1;
+        case '+':
+            break;
+        default:
+            return -5;
+    }
+
+parse_timezone:;
```

Comment:
Unfortunately yes, there is. It's a syntax error without the `;`, because the following line is a variable declaration, not a statement, the semicolon is there to provide an empty statement for the `goto` to go to.

That said, there's no particular reason why those variables can't be declared at the top of the file (before the first call to `PARSE_ISOFORMAT_ADVANCE_TIME_SEP`).